### PR TITLE
Add Support GitHub Enterprise Server URL pattern

### LINKED ISSUE #1268: Add support for GitHub Enterprise
## Description
When the provider is set to 'github', the current implementation fails to correctly parse URLs from GitHub Enterprise Server instances. This causes errors when users attempt to use Enterprise Server URLs with our system.

## Current Behavior
- URLs from github.com are parsed correctly
- URLs from GitHub Enterprise Server instances throw an error or are parsed incorrectly

## Expected Behavior
- Both github.com and GitHub Enterprise Server URLs should be parsed correctly when the provider is set to 'github'

## Steps to Reproduce
1. Set the provider to 'github'
2. Attempt to use a URL from a GitHub Enterprise Server instance (e.g., https://github.mycompany.com/org/repo/pull/123)
3. Observe the error or incorrect parsing

## Proposed Solution
Update the URL parsing logic to accommodate both github.com and GitHub Enterprise Server URL structures. This may involve:
- Using more flexible regex patterns
- Adjusting the logic to handle varying URL formats
- Ensuring that repository names and PR/issue numbers are correctly extracted regardless of the exact URL structure

## Additional Notes
This issue is related to the recent changes made in the PR and issue URL parsing functions. While those changes improved the situation, they didn't fully address the Enterprise Server use case.

### PR DESCRIPTION
**Pull request by yamao-latte**
### **User description**
- Update _parse_pr_url and _parse_issue_url functions to use regex
- Improve handling of various GitHub URL formats, including Enterprise Server
- Simplify logic and reduce code duplication
- Enhance error handling for invalid URL structures

Closes #1268 

## Description
When the provider is set to 'github', the current implementation fails to correctly parse URLs from GitHub Enterprise Server instances. This causes errors when users attempt to use Enterprise Server URLs with our system.

## Current Behavior
- URLs from github.com are parsed correctly
- URLs from GitHub Enterprise Server instances throw an error or are parsed incorrectly

## Expected Behavior
- Both github.com and GitHub Enterprise Server URLs should be parsed correctly when the provider is set to 'github'

## Steps to Reproduce
1. Set the provider to 'github'
2. Attempt to use a URL from a GitHub Enterprise Server instance (e.g., https://github.mycompany.com/org/repo/pull/123)
3. Observe the error or incorrect parsing

## Proposed Solution
Update the URL parsing logic to accommodate both github.com and GitHub Enterprise Server URL structures. This may involve:
- Using more flexible regex patterns
- Adjusting the logic to handle varying URL formats
- Ensuring that repository names and PR/issue numbers are correctly extracted regardless of the exact URL structure

## Additional Notes
This issue is related to the recent changes made in the PR and issue URL parsing functions. While those changes improved the situation, they didn't fully address the Enterprise Server use case.

___

### **PR Type**
Enhancement

___

### **Description**
- Implemented regex-based parsing for PR and issue URLs, improving compatibility with various GitHub URL formats, including Enterprise Server
- Updated base_url_html logic to handle different GitHub URL structures (cloud, enterprise with api/v3, and others)
- Refactored _parse_pr_url and _parse_issue_url functions to use regex patterns, simplifying logic and reducing code duplication
- Enhanced error handling for invalid URL structures
- Improved overall flexibility and robustness of URL parsing for both github.com and GitHub Enterprise Server instances

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>github_provider.py</strong><dd><code>Enhance GitHub URL parsing and compatibility</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

pr_agent/git_providers/github_provider.py

<li>Added regex import and implemented regex-based URL parsing for PRs and <br>issues<br> <li> Updated base_url_html logic to handle different GitHub URL formats<br> <li> Refactored _parse_pr_url and _parse_issue_url functions for improved <br>compatibility<br> <li> Simplified error handling and reduced code duplication<br>

</details>

  </td>
  <td><a href="https://github.com/Codium-ai/pr-agent/pull/1270/files#diff-28979ec713529d6f619bcba0b9e6a3002d1164d47c70ccfea7f15306618a1a11">+26/-34</a>&nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

**Comment by mrT23** [comment]
As this PR changes a critical component, the level of tests and code quality needed to approve this PR is high.

As a general rule of thumb, as few changes as possible is the best way to go.

I will also give code comments, but comprehensive **tests**, to ensure that compatibility to the GitHub cloud remains, are a must to this PR be considered to merge.

File: pr_agent/git_providers/github_provider.py [hunk 1]
```diff
@@ -28,7 +29,14 @@ def __init__(self, pr_url: Optional[str] = None):
...
         self.base_url = get_settings().get("GITHUB.BASE_URL", "https://api.github.com").rstrip("/")
-        self.base_url_html = self.base_url.split("api/")[0].rstrip("/") if "api/" in self.base_url else "https://github.com"
+        if self.base_url.endswith("/api/v3"):
+            # enterprise server with api/v3
+            self.base_url_html = self.base_url[:len("/api/v3")]
+        elif self.base_url == "https://api.github.com":
+            # github cloud
+            self.base_url_html = "https://github.com"
+        else:
+            self.base_url_html = self.base_url
```
Comment (mrT23) [inline pr_agent/git_providers/github_provider.py:39]: explain this change.
why is it needed ?

Give concrete examples - what didn't work before, and is working now due to this change ?

File: pr_agent/git_providers/github_provider.py [hunk 2]
```diff
@@ -611,51 +619,35 @@ def remove_reaction(self, issue_comment_id: int, reaction_id: str) -> bool:
...
-
-        repo_name = '/'.join(path_parts[:2])
-        try:
-            issue_number = int(path_parts[3])
-        except ValueError as e:
-            raise ValueError("Unable to convert issue number to integer") from e
-
-        return repo_name, issue_number
+        else:
+            raise ValueError("The provided URL does not appear to be a GitHub issue URL")
```
Comment (mrT23) [inline pr_agent/git_providers/github_provider.py:651]: does the new code supports:
'https://github.com/Codium-ai/pr-agent/issues/1268'
?

File: pr_agent/git_providers/github_provider.py [hunk 3]
```diff
@@ -611,51 +619,35 @@ def remove_reaction(self, issue_comment_id: int, reaction_id: str) -> bool:
...
     def _parse_issue_url(self, issue_url: str) -> Tuple[str, int]:
         parsed_url = urlparse(issue_url)
-        path_parts = parsed_url.path.strip('/').split('/')
-        if self.base_domain in parsed_url.netloc:
-            if len(path_parts) < 5 or path_parts[3] != 'issues':
-                raise ValueError("The provided URL does not appear to be a GitHub ISSUE URL")
-            repo_name = '/'.join(path_parts[1:3])
+        path = parsed_url.path
+        
+        pattern = r'/repos/([^/]+/[^/]+)/issues/(\d+)'
```
Comment (mrT23) [inline pr_agent/git_providers/github_provider.py:639]: would this format support:

"https://github.com/Codium-ai/pr-agent/issues/1268"

for example
??

File: pr_agent/git_providers/github_provider.py [hunk 4]
```diff
@@ -611,51 +619,35 @@ def remove_reaction(self, issue_comment_id: int, reaction_id: str) -> bool:
     def _parse_pr_url(self, pr_url: str) -> Tuple[str, int]:
         parsed_url = urlparse(pr_url)
 
-        path_parts = parsed_url.path.strip('/').split('/')
-        if self.base_domain in parsed_url.netloc:
-            if len(path_parts) < 5 or path_parts[3] != 'pulls':
-                raise ValueError("The provided URL does not appear to be a GitHub PR URL")
-            repo_name = '/'.join(path_parts[1:3])
+        pattern = r'/repos/([^/]+/[^/]+)/pulls?/(\d+)'
```
Comment (mrT23) [inline pr_agent/git_providers/github_provider.py:622]: would this format support

`https://github.com/Codium-ai/pr-agent/pull/1270`

for example ??

Comment (yamao-latte) [reply pr_agent/git_providers/github_provider.py:622]: We would like to support a format such as:
`https://github.enterprise-server.com/api/v3/Codium-ai/pr-agent/pull/1270`

In GitHub Enterprise Server, the API URLs include the `api/v3` segment. The current implementation uses a hard-coded check:

```
if len(path_parts) < 5 or path_parts[3] != 'pulls':
```
This approach does not account for the `api/v3` part of the URL, leading to errors when the URL structure differs. Specifically, this issue affects the following part of the code:

For more information on the URL structure, refer to the GitHub Enterprise Server documentation: 
[GitHub Docs - Get a Pull Request](https://docs.github.com/ja/enterprise-server@3.14/rest/pulls/pulls?apiVersion=2022-11-28#get-a-pull-request)

https://github.com/Codium-ai/pr-agent/blob/ada0a3d10f357e18033b4f6772eac35b79a2c369/pr_agent/servers/github_action_runner.py#L91-L115

File: pr_agent/git_providers/github_provider.py (hunk 3 above)
Comment (yamao-latte) [reply pr_agent/git_providers/github_provider.py:639]: We would like to support a format such as:
`https://github.enterprise-server.com/api/v3/Codium-ai/pr-agent/issues/1268`

In GitHub Enterprise Server, the API URLs include the `api/v3` segment. The current implementation uses a hard-coded check:

```
            if len(path_parts) < 5 or path_parts[3] != 'issues':
```
This approach does not account for the `api/v3` part of the URL, leading to errors when the URL structure differs. Specifically, this issue affects the following part of the code:

For more information on the URL structure, refer to the GitHub Enterprise Server documentation: 
[GitHub Docs - Get a Pull Request](https://docs.github.com/ja/enterprise-server@3.14/rest/pulls/pulls?apiVersion=2022-11-28#get-a-pull-request)

**Comment by mrT23** [comment]
@yamao-latte 
If you are not going to address and improve this PR, I suggest closing it.
It cannot be merged in its current form

You can still utilize the code in your local deployment

File: pr_agent/git_providers/github_provider.py [hunk 5]
```diff
@@ -28,7 +29,14 @@ def __init__(self, pr_url: Optional[str] = None):
             self.installation_id = None
         self.max_comment_chars = 65000
         self.base_url = get_settings().get("GITHUB.BASE_URL", "https://api.github.com").rstrip("/")
-        self.base_url_html = self.base_url.split("api/")[0].rstrip("/") if "api/" in self.base_url else "https://github.com"
+        if self.base_url.endswith("/api/v3"):
```
Comment (mrT23) [inline pr_agent/git_providers/github_provider.py:32]:
```suggestion
        if self.base_url.endswith("/api/v4"):
```

**Comment by mrT23** [comment]
anyway,  github serve support was added here
https://qodo-merge-docs.qodo.ai/installation/github/#action-for-github-enterprise-server

you are welcome to try it and give feedback

**Pull request closed without merging**